### PR TITLE
[DONTMERGE] check: make sure all files outside contrib/ include "config.h" first.

### PR DESCRIPTION
**Pull request by rustyrussell**
And turn "" includes into full-path (which makes it easier to put
config.h first, and finds some cases check-includes.sh missed
previously).

config.h sets _GNU_SOURCE which really needs to be done before any
'#includes': we mainly got away with it with glibc, but other platforms
like Alpine may have stricter requirements.

Signed-off-by: Rusty Russell <rusty@rustcorp.com.au>

### CODE REVIEW

File: ccan_compat.h [hunk 1]
```diff
@@ -1,7 +1,12 @@
 #ifndef LIGHTNING_CCAN_COMPAT_H
 #define LIGHTNING_CCAN_COMPAT_H
+
 /* Magical file included from config.h (ie. everywhere) which renames
  * sha256 routines so they don't clash with libwally-core's internal ones */
+
+/* So, for obvious reasons, this is an exception to the usual rule that we
+#include "config.h"
```
Comment (ZmnSCPxj) [inline ccan_compat.h:8]: Well, include guards protect against circular inclusions...

On the other hand, yeah, an exception for this is totally better.

**Review by ZmnSCPxj — APPROVED** [review]
ACK 460b72435c3109143110b0c4a8b5249ff55aec1c

**Review by vincenzopalazzo — CHANGES_REQUESTED** [review]
During my test on alpine, I got the following error

```
cc common/bolt11_json.c
cc common/bolt12.c
cc common/configdir.c
cc common/status_levels.c
cc common/io_lock.c
In file included from common/io_lock.c:3:
ccan/ccan/io/io_plan.h:13:2: error: unknown type name 'size_t'
   13 |  size_t s;
      |  ^~~~~~
ccan/ccan/io/io_plan.h:14:16: error: 'size_t' undeclared here (not in a function)
   14 |  char c[sizeof(size_t)];
      |                ^~~~~~
ccan/ccan/io/io_plan.h:1:1: note: 'size_t' is defined in header '<stddef.h>'; did you forget to '#include <stddef.h>'?
  +++ |+#include <stddef.h>
    1 | /* Licensed under LGPLv2.1+ - see LICENSE file for details */
make: *** [Makefile:279: common/io_lock.o] Error 1
make: *** Waiting for unfinished jobs....
rm external/x86_64-alpine-linux-musl/libwally-core-build/src/secp256k1/libsecp256k1.la

```
Maybe we ~need to add also the config header before the comm/io_lock.c file or~ follow the suggestion of the compiler? 

P.S: Not sure to have understood why the error happens

**Comment by rustyrussell** [comment]
Ok, we can actually remove that file as it's no longer used.

**Comment by rustyrussell** [comment]
Hmm, @ProofOfKeags on Telegram says it still doesn't work for him (still has a bunch of fortify errors).

@jsarenik any chance you can take a look?

**Comment by ProofOfKeags** [comment]
I can provide any details needed as well. Here's the original error which persists even after some of the changes in this PR:

```
#18 362.1 In file included from /usr/include/fortify/strings.h:22,
#18 362.1                  from /usr/include/string.h:57,
#18 362.1                  from /usr/include/fortify/string.h:22,
#18 362.1                  from ccan/ccan/str/str.h:5,
#18 362.1                  from ccan/ccan/tal/tal.h:8,
#18 362.1                  from ./common/amount.h:5,
#18 362.1                  from cli/test/run-human-mode.c:2:
#18 362.1 /usr/include/fortify/unistd.h:110:1: error: 'read' undeclared here (not in a function); did you mean 'pread'?
#18 362.1   110 | _FORTIFY_FN(read) ssize_t read(int __f, void *__s, size_t __n)
#18 362.1       | ^~~~~~~~~~~
#18 362.1 In file included from cli/test/../lightning-cli.c:22,
#18 362.1                  from cli/test/run-human-mode.c:27:
#18 362.1 /usr/include/fortify/unistd.h: In function 'read':
#18 362.1 /usr/include/fortify/unistd.h:116:9: error: called object '__orig_read' is not a function or function pointer
#18 362.1   116 |  return __orig_read(__f, __s, __n);
#18 362.1       |         ^~~~~~~~~~~
#18 362.1 In file included from /usr/include/fortify/strings.h:22,
#18 362.1                  from /usr/include/string.h:57,
#18 362.1                  from /usr/include/fortify/string.h:22,
#18 362.1                  from ccan/ccan/str/str.h:5,
#18 362.1                  from ccan/ccan/tal/tal.h:8,
#18 362.1                  from ./common/amount.h:5,
#18 362.1                  from cli/test/run-human-mode.c:2:
#18 362.1 /usr/include/fortify/unistd.h:110:1: note: declared here
#18 362.1   110 | _FORTIFY_FN(read) ssize_t read(int __f, void *__s, size_t __n)
#18 362.1       | ^~~~~~~~~~~
#18 362.4 In file included from /usr/include/fortify/strings.h:22,
#18 362.4                  from /usr/include/string.h:57,
#18 362.4                  from /usr/include/fortify/string.h:22,
#18 362.4                  from ccan/ccan/str/str.h:5,
#18 362.4                  from ccan/ccan/tal/tal.h:8,
#18 362.4                  from ./common/amount.h:5,
#18 362.4                  from cli/test/run-large-input.c:2:
#18 362.4 /usr/include/fortify/unistd.h:110:1: error: 'read' undeclared here (not in a function); did you mean 'pread'?
#18 362.4   110 | _FORTIFY_FN(read) ssize_t read(int __f, void *__s, size_t __n)
#18 362.4       | ^~~~~~~~~~~
#18 362.4 In file included from cli/test/../lightning-cli.c:22,
#18 362.4                  from cli/test/run-large-input.c:27:
#18 362.4 /usr/include/fortify/unistd.h: In function 'read':
#18 362.4 /usr/include/fortify/unistd.h:116:9: error: called object '__orig_read' is not a function or function pointer
#18 362.4   116 |  return __orig_read(__f, __s, __n);
#18 362.4       |         ^~~~~~~~~~~
#18 362.4 In file included from /usr/include/fortify/strings.h:22,
#18 362.4                  from /usr/include/string.h:57,
#18 362.4                  from /usr/include/fortify/string.h:22,
#18 362.4                  from ccan/ccan/str/str.h:5,
#18 362.4                  from ccan/ccan/tal/tal.h:8,
#18 362.4                  from ./common/amount.h:5,
#18 362.4                  from cli/test/run-large-input.c:2:
#18 362.4 /usr/include/fortify/unistd.h:110:1: note: declared here
#18 362.4   110 | _FORTIFY_FN(read) ssize_t read(int __f, void *__s, size_t __n)
#18 362.4       | ^~~~~~~~~~~
#18 362.5 make: *** [Makefile:279: cli/test/run-human-mode.o] Error 1
#18 362.5 make: *** Waiting for unfinished jobs....
#18 362.8 make: *** [Makefile:279: cli/test/run-large-input.o] Error 1
#18 363.1 In file included from /usr/include/fortify/strings.h:22,
#18 363.1                  from /usr/include/string.h:57,
#18 363.1                  from /usr/include/fortify/string.h:22,
#18 363.1                  from ccan/ccan/str/str.h:5,
#18 363.1                  from ccan/ccan/tal/tal.h:8,
#18 363.1                  from ./common/amount.h:5,
#18 363.1                  from cli/test/run-remove-hint.c:2:
#18 363.1 /usr/include/fortify/unistd.h:110:1: error: 'read' undeclared here (not in a function); did you mean 'pread'?
#18 363.1   110 | _FORTIFY_FN(read) ssize_t read(int __f, void *__s, size_t __n)
#18 363.1       | ^~~~~~~~~~~
#18 363.1 In file included from cli/test/../lightning-cli.c:22,
#18 363.1                  from cli/test/run-remove-hint.c:30:
#18 363.1 /usr/include/fortify/unistd.h: In function 'read':
#18 363.1 /usr/include/fortify/unistd.h:116:9: error: called object '__orig_read' is not a function or function pointer
#18 363.1   116 |  return __orig_read(__f, __s, __n);
#18 363.1       |         ^~~~~~~~~~~
#18 363.1 In file included from /usr/include/fortify/strings.h:22,
#18 363.1                  from /usr/include/string.h:57,
#18 363.1                  from /usr/include/fortify/string.h:22,
#18 363.1                  from ccan/ccan/str/str.h:5,
#18 363.1                  from ccan/ccan/tal/tal.h:8,
#18 363.1                  from ./common/amount.h:5,
#18 363.1                  from cli/test/run-remove-hint.c:2:
#18 363.1 /usr/include/fortify/unistd.h:110:1: note: declared here
#18 363.1   110 | _FORTIFY_FN(read) ssize_t read(int __f, void *__s, size_t __n)
#18 363.1       | ^~~~~~~~~~~
#18 363.5 make: *** [Makefile:279: cli/test/run-remove-hint.o] Error 1
#18 379.1 rm external/armv7-alpine-linux-musleabihf/libwally-core-build/src/secp256k1/libsecp256k1.la
```

**Comment by jsarenik** [comment]
That is interesting. Package `fortify-headers` on Alpine is installed as dependency of `build-base`. Sniffing, as I just had a similar fortify-related (but just a warning) [on Bitcoin Core](https://github.com/bitcoin/bitcoin/issues/23638). But it seems to me that `fortify-headers` is not really needed, just causing troubles. Double-checking now on Alpine systems without that package installed.

**Comment by chrisguida** [comment]
Is there a workaround for this? Builds on alpine still failing for me :/

https://github.com/ElementsProject/lightning/pull/4950#issuecomment-985907455

**Comment by vincenzopalazzo** [comment]
@chrisguida 

`rm -r cli/test/*.c ` in the root of lightning

**Comment by chrisguida** [comment]
Works! Thanks, @vincenzopalazzo!!

**Comment by cdecker** [comment]
Closing this for now as it causes conflicts in many files it touched, and there is no plan to merge it as is.

**Pull request closed without merging**